cmake/kokkos_tribits.cmake

```diff
@@ -0,0 +1,463 @@
+  ENDIF()
+  IF(KOKKOS_XCOMPILER_OPTIONS)
+    SET(NODEDUP_XCOMPILER_OPTIONS)
+    FOREACH(OPT ${KOKKOS_XCOMPILER_OPTIONS})
+      #I have to do this for now because we can't guarantee 3.12 support
+      #I really should do this with the shell option 
+      LIST(APPEND NODEDUP_XCOMPILER_OPTIONS -Xcompiler) 
+      LIST(APPEND NODEDUP_XCOMPILER_OPTIONS ${OPT}) 
+    ENDFOREACH()
+    TARGET_COMPILE_OPTIONS(
+      ${LIBRARY_NAME} 
+      PUBLIC $<$<COMPILE_LANGUAGE:CXX>:${NODEDUP_XCOMPILER_OPTIONS}>
+    )
+  ENDIF()
+
+
+
+  TARGET_INCLUDE_DIRECTORIES(
+    ${LIBRARY_NAME}
+    PUBLIC ${KOKKOS_TPL_INCLUDE_DIRS}
+  )
+
+  IF (KOKKOS_ENABLE_CUDA)
+    SET(LIB_cuda "-lcuda")
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC cuda)
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_HPX)
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC ${HPX_LIBRARIES})
+    TARGET_INCLUDE_DIRECTORIES(${LIBRARY_NAME} PUBLIC ${HPX_INCLUDE_DIRS})
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_HWLOC)
+    #this is really annoying that I have to do this
+    #if CMake links statically, it will not link in hwloc which causes undefined refs downstream
+    #even though hwloc is really "private" and doesn't need to be public I have to link publicly
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC hwloc)
+    #what I don't want is the headers to be propagated downstream
+    TARGET_INCLUDE_DIRECTORIES(${LIBRARY_NAME} PRIVATE ${HWLOC_INCLUDE_DIR})
+  ENDIF()
+  
+  IF (KOKKOS_ENABLE_LIBNUMA)
+    #this is really annoying that I have to do this
+    #if CMake links statically, it will not link in hwloc which causes undefined refs downstream
+    #even though hwloc is really "private" and doesn't need to be public I have to link publicly
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC libnuma)
+    #what I don't want is the headers to be propagated downstream
+    TARGET_INCLUDE_DIRECTORIES(${LIBRARY_NAME} PRIVATE ${LIBNUMA_INCLUDE_DIR})
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_LIBRT)
+    #this is really annoying that I have to do this
+    #if CMake links statically, it will not link in librt which causes undefined refs downstream
+    #even though librt is really "private" and doesn't need to be public I have to link publicly
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PRIVATE librt)
+    #what I don't want is the headers to be propagated downstream
+    TARGET_INCLUDE_DIRECTORIES(${LIBRARY_NAME} PRIVATE ${LIBRT_INCLUDE_DIR})
+  ENDIF()
+
+  IF (KOKKOS_ENABLE_MEMKIND)
+    TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PRIVATE memkind)
+  ENDIF()
+
+  IF (KOKKOS_CXX_STANDARD_FEATURE)
+    #GREAT! I can't do this the right way
+    TARGET_COMPILE_FEATURES(${LIBRARY_NAME} PUBLIC ${KOKKOS_CXX_STANDARD_FEATURE})
+  ELSE()
+    #OH, Well, no choice but the wrong way
+    TARGET_COMPILE_OPTIONS(${LIBRARY_NAME} PUBLIC ${KOKKOS_CXX_STANDARD_FLAG})
+  ENDIF()
+
+  #dlfcn.h is in header files and needs to propagate
+  TARGET_LINK_LIBRARIES(${LIBRARY_NAME} PUBLIC libdl)
```

Comment:
We are getting `/usr/bin/ld: cannot find -llibdl` because "libdl" is interpreted as a library name and added to the link line (instead of being interpreted as a target)

You should namespace TPLs targets.

Reply:
I did the namespacing and it caused issues downstream. I don't know why. It should have worked. This should have crashed before you got to this error. The `find_package` call should force you to find libdl and it should always be a valid target.

Reply:
What kind of problems did you see downstream?

Reply:
I honestly don't remember exactly. Had to do with the way I was trying to "export" the imported target. @dalg24 - you must have been getting this in a downstream project, not Kokkos itself or the tests?

I was missing the re-find modules in the KokkosConfig.cmake. I changed the way things get re-imported in KokkosConfig.cmake and am now using namespaces. Hopefully this fixes your issues.